From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: christophert <christophert@nvidia.com>
Date: Tue, 9 Dec 2025 17:43:27 +0000
Subject: [PATCH 116/243] NVIDIA: SAUCE: virt: tegra: add nvlog ability to
 access hypervisor from sysfs

Added NvLog Entity sysfs support to expose nvlog_entities[] from the
hypervisor's vm_info_region to userspace.

Purpose
This allows Linux userspace (nvlog_gos_impl_linux.c and
nvlog_llc_table_os_linux.c) to:
Read entity information from sysfs (instead of making direct hypervisor
calls)
Mmap the LLC table for live log level control

Jira HYP-23140

Bug 5658048

Change-Id: Iefd57e00d21ee3555729cef0a89e550f759d00eb
Signed-off-by: christophert <christophert@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3532106
Reviewed-by: Manish Bhardwaj <mbhardwaj@nvidia.com>
Tested-by: Manish Bhardwaj <mbhardwaj@nvidia.com>
Reviewed-by: Brad Griffis <bgriffis@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
(cherry picked from commit 9c721fdd8b2412f407b9ef9c5847bd5bcb67c237)

diff --git a/include/soc/tegra/virt/syscalls.h b/include/soc/tegra/virt/syscalls.h
index bb9203945305..e087b25173aa 100644
--- a/include/soc/tegra/virt/syscalls.h
+++ b/include/soc/tegra/virt/syscalls.h
@@ -1,6 +1,6 @@
 /*
  * SPDX-License-Identifier: GPL-2.0-only
- * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  */
 
 #ifndef TEGRA_SYSCALLS_H
@@ -65,6 +65,8 @@
 #define GUEST_IVC_SERVER	0
 /* @brief Max number of nvlog producers */
 #define MAX_NVLOG_PRODUCERS	32U
+/* @brief Max number of nvlog entities */
+#define MAX_NVLOG_ENTITIES	3U
 /* @brief Immediate hypercall value used in
  * - hyp_read_freq_feedback() wrapper
  * - hyp_read_freq_request() wrapper
@@ -116,6 +118,20 @@ struct nvlog_producer {
 	char name[32];
 };
 
+/*
+ * @brief Structure to store information about an NvLog entity. An entity is
+ * generally an operating system deployment. Log producers belonging to an
+ * entity will inhabit the same PID space and share a log level region.
+ */
+struct nvlog_entity {
+	/* @brief The NvLog entity ID. */
+	uint16_t id;
+	/* @brief IPA of shared memory used to request NvLog log level changes. */
+	uint64_t log_level_region_ipa;
+	/* @brief Size of shared memory used to request NvLog log level changes. */
+	uint64_t log_level_region_size;
+} __attribute__((packed, aligned(8)));
+
 /* @brief
  * Data structure for the VM Information Region.
  *
@@ -131,6 +147,11 @@ struct vm_info_region {
 	 * and buf_count' fields.
 	 */
 	struct nvlog_producer nvlog_producers[MAX_NVLOG_PRODUCERS];
+
+	/**
+	 * @brief NvLog Entity Info.
+	 */
+	struct nvlog_entity nvlog_entities[MAX_NVLOG_ENTITIES];
 };
 
 /**
-- 
2.43.0

